src/analyses/cfg_dominators.h

```diff
@@ -44,143 +246,352 @@ class cfg_dominators_templatet
 
 protected:
   void initialise(P &program);
-  void fixedpoint(P &program);
-};
 
-/// Print the result of the dominator computation
-template <class P, class T, bool post_dom>
-std::ostream &operator << (
-  std::ostream &out,
-  const cfg_dominators_templatet<P, T, post_dom> &cfg_dominators)
-{
-  cfg_dominators.output(out);
-  return out;
-}
+  struct fixedpointt
+  {
+    explicit fixedpointt(cfg_dominators_templatet &cfg_dominators)
+      : cfg_dominators(cfg_dominators),
+        dfs_counter(0),
+        // Data structures have size cfg.size() + 1 as node indices are 1-based
```

Comment:
Is it important that the reserved 0 is lower than all the allocated indices (e.g. so that the null node <= the root which is <= all other nodes?) If not suggest just using 0-based indices and note the deviation from the paper.

Reply:
@smowton I think not, IIRC I just got fed up trying to translate the logic from the paper and eventually just opted to just replicate their logic exactly, including indexing.